libbeat/tests/integration/elasticsearch_test.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/rcrowley/go-metrics"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/mock-es/pkg/api"
+)
+
+var esCfg = `
+mockbeat:
+logging:
+  level: debug
+  selectors:
+    - publisher_pipeline_output
+    - esclientleg
+queue.mem:
+  events: 4096
+  flush.min_events: 8
+  flush.timeout: 0.1s
+output.elasticsearch:
+  allow_older_versions: true
+  hosts:
+    - "http://localhost:4242"
+  backoff:
+    init: 0.1s
+    max: 0.2s
+`
+
+func TestESOutputRecoversFromNetworkError(t *testing.T) {
+	mockbeat := NewBeat(t, "mockbeat", "../../libbeat.test")
+	mockbeat.WriteConfigFile(esCfg)
+
+	s, mr := startMockES(t, "localhost:4242")
+
+	mockbeat.Start()
+
+	// 1. Wait for one _bulk call
+	waitForEventToBePublished(t, mr)
+
+	// 2. Stop the mock-es server
+	if err := s.Close(); err != nil {
+		t.Fatalf("cannot close mock-es server: %s", err)
+	}
+
+	// 3. Wait for connection error logs
+	mockbeat.WaitForLogs(
+		`Get \"http://localhost:4242\": dial tcp 127.0.0.1:4242: connect: connection refused`,
+		2*time.Second,
+		"did not find connection refused error")
+
+	mockbeat.WaitForLogs(
+		"Attempting to reconnect to backoff(elasticsearch(http://localhost:4242)) with 2 reconnect attempt(s)",
+		2*time.Second,
+		"did not find two tries to reconnect")
+
+	// 4. Restart mock-es on the same port
+	s, mr = startMockES(t, "localhost:4242")
+
+	// 5. Wait for reconnection logs
+	mockbeat.WaitForLogs(
+		"Connection to backoff(elasticsearch(http://localhost:4242)) established",
+		5*time.Second, // There is a backoff, so ensure we wait enough
+		"did not find re connection confirmation")
+
+	// 6. Ensure one new call to _bulk is made
+	waitForEventToBePublished(t, mr)
+	s.Close()
+}
+
+func startMockES(t *testing.T, addr string) (*http.Server, metrics.Registry) {
+	uid := uuid.Must(uuid.NewV4())
+	mr := metrics.NewRegistry()
+	es := api.NewAPIHandler(uid, "foo2", mr, time.Now().Add(24*time.Hour), 0, 0, 0, 0, 0)
+
+	s := http.Server{Addr: addr, Handler: es, ReadHeaderTimeout: time.Second}
+	go func() {
+		if err := s.ListenAndServe(); !errors.Is(http.ErrServerClosed, err) {
+			t.Errorf("could not start mock-es server: %s", err)
+		}
+	}()
+
+	require.Eventually(t, func() bool {
+		resp, err := http.Get("http://" + addr) //nolint: noctx // It's just a test
+		if err != nil {
+			//nolint: errcheck // We're just draining the body, we can ignore the error
+			io.Copy(io.Discard, resp.Body)
+			resp.Body.Close()
+			return false
+		}
+		return true
+	},
+		time.Second, time.Millisecond, "first server must be up")
+
+	return &s, mr
+}
+
+func waitForEventToBePublished(t *testing.T, mr metrics.Registry) {
+	t.Helper()
+	require.Eventually(t, func() bool {
+		total := mr.Get("bulk.create.total")
+		if total == nil {
+			return false
+		}
+
+		sc, ok := total.(*metrics.StandardCounter)
+		if !ok {
+			t.Fatalf("expecting 'bulk.create.total' to be *metrics.StandardCounter, but got '%T' instead",
+				total,
+			)
+		}
+
+		return sc.Count() > 1
+	},
```

Comment:
So, doubt I have, how do you ensure the counter does not have a "old" request, as if it happened before the test starts "waiting" for the request to be made?

Reply:
The metrics registry is passed as an argument, so it's the caller's responsibility to ensure the metrics are "clean".

I'll add some docs better explaining it.